cloudwash/entities/providers.py

```diff
@@ -0,0 +1,37 @@
+from cloudwash.entities.resources.discs import CleanAWSDiscs
+from cloudwash.entities.resources.discs import CleanAzureDiscs
+from cloudwash.entities.resources.vms import CleanAWSVms
+from cloudwash.entities.resources.vms import CleanAzureVMs
+
+
+class providerCleanup:
+    def __init__(self, client):
+        self.client = client
+
+    @property
+    def vms(self):
+        providerclass = self.__class__.__name__
+        if 'Azure' in providerclass:
+            return CleanAzureVMs(client=self.client)
+        elif 'AWS' in providerclass:
+            return CleanAWSVms(client=self.client)
```

Comment:
@jyejare one more approach to solve this to remove more coupling and adding more providers and reduce the factory pattern made in parent class. let me know what you think about it  

``` python 
class ProviderCleanup:
    def __init__(self, client, vm_cleaner, disc_cleaner):
        self.client = client
        self.vm_cleaner = vm_cleaner
        self.disc_cleaner = disc_cleaner

    def vms(self):
        return self.vm_cleaner.clean(self.client)

    def discs(self):
        return self.disc_cleaner.clean(self.client)


class AzureCleanup(ProviderCleanup):
    def __init__(self, client):
        super().__init__(client, CleanAzureVMs(), CleanAzureDiscs())


class AWSCleanup(ProviderCleanup):
    def __init__(self, client):
        super().__init__(client, CleanAWSVms(), CleanAWSDiscs())
```

Reply:
Hmm! 

I see multiple issues in this pattern:
1. Unnecessary Object Initialization of CloudProviderResourceType classes (e.g CleanAzureVMs(client), CleanAzureDiscs(client)) even though its not demanded! That also takes a huge time for all classes initialization. For now its the object creation is on demand and explicit if its demanded! 
2. The list of parameters will increase in endless number as and when number of resource types will grow. Though we can control it with kwargs, but its extra layer of verification if the object is available in params.
3. In current design, if the non-common resource type property is needed to be added then you can do it at that cloud provider class level and dont need to send it back to base class for processing / returning.

For now, I can only think of these issues! Feel free to reply with the thoughts !

Reply:
Okay, I ack it.